staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go

```diff
@@ -17,327 +17,17 @@ limitations under the License.
 package feature
 
 import (
-	"fmt"
-	"sort"
-	"strconv"
-	"strings"
-	"sync"
-	"sync/atomic"
-
-	"github.com/spf13/pflag"
-	"k8s.io/klog"
-)
-
-type Feature string
-
-const (
-	flagName = "feature-gates"
-
-	// allAlphaGate is a global toggle for alpha features. Per-feature key
-	// values override the default set by allAlphaGate. Examples:
-	//   AllAlpha=false,NewFeature=true  will result in newFeature=true
-	//   AllAlpha=true,NewFeature=false  will result in newFeature=false
-	allAlphaGate Feature = "AllAlpha"
+	"k8s.io/component-base/featuregate"
 )
 
 var (
-	// The generic features.
-	defaultFeatures = map[Feature]FeatureSpec{
-		allAlphaGate: {Default: false, PreRelease: Alpha},
-	}
-
-	// Special handling for a few gates.
-	specialFeatures = map[Feature]func(known map[Feature]FeatureSpec, enabled map[Feature]bool, val bool){
-		allAlphaGate: setUnsetAlphaGates,
-	}
-
 	// DefaultMutableFeatureGate is a mutable version of DefaultFeatureGate.
-	// Only top-level commands/options setup and the k8s.io/apiserver/pkg/util/feature/testing package should make use of this.
+	// Only top-level commands/options setup and the k8s.io/component-base/featuregate/testing package should make use of this.
 	// Tests that need to modify feature gates for the duration of their test should use:
-	//   defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()
-	DefaultMutableFeatureGate MutableFeatureGate = NewFeatureGate()
+	//   defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)()
+	DefaultMutableFeatureGate featuregate.MutableFeatureGate = featuregate.NewFeatureGate()
 
 	// DefaultFeatureGate is a shared global FeatureGate.
 	// Top-level commands/options setup that needs to modify this feature gate should use DefaultMutableFeatureGate.
-	DefaultFeatureGate FeatureGate = DefaultMutableFeatureGate
-)
-
-type FeatureSpec struct {
-	// Default is the default enablement state for the feature
-	Default bool
-	// LockToDefault indicates that the feature is locked to its default and cannot be changed
-	LockToDefault bool
-	// PreRelease indicates the maturity level of the feature
-	PreRelease prerelease
-}
-
-type prerelease string
-
-const (
-	// Values for PreRelease.
-	Alpha = prerelease("ALPHA")
-	Beta  = prerelease("BETA")
-	GA    = prerelease("")
-
-	// Deprecated
-	Deprecated = prerelease("DEPRECATED")
+	DefaultFeatureGate featuregate.FeatureGate = DefaultMutableFeatureGate
```

Comment:
Can we move this to component-base with the rest of the feature gate code? There are several places we use this outside of the apiserver including some controllers and the gce_disks portions of the gce cloud provider. I feel it would be cleaner to keep the feature gate code in one place and minimize the dependencies which are pulled in when accessing features.

Reply:
hmm, see discussion at https://github.com/kubernetes/kubernetes/pull/74021#discussion_r257702497

Reply:
There was a comment on this previously https://github.com/kubernetes/kubernetes/pull/74021#discussion_r257702497. The `DefaultFeatureGate` is a global state that currently (sort of?) belongs to the apiserver. I think it may not make much sense to move it into `k8s.io/component-base` since what we add there may not apply to all components in the future. I don't have strong opinions on this though and happy to move it if it makes sense

Reply:
https://github.com/kubernetes/kubernetes/pull/77412 removed the need to do this for gce_disks' benefit